### PR TITLE
Get credentials vm

### LINKED ISSUE #19: The csp-billing-adapter-microsoft needs to support Azure Managed applications
A few difference exist in how the token, resourceURI and plan are collected from a Azure Manager Kubernetes Application and a standard  Azure Managed  (VM).

In order for the token to be retrieved, the VM must have Azure System Identity assigned

https://learn.microsoft.com/en-us/partner-center/marketplace/marketplace-metering-service-authentication#using-the-azure-managed-identities-token

1) For the Managed Application, the token needs to be acquired by using `resource=https://management.azure.com`  and not needing a client id https://github.com/SUSE-Enceladus/csp-billing-adapter-microsoft/blob/main/csp_billing_adapter_microsoft/plugin.py#L211

2) I believe the method for getting the planID can leverage the metadata as it has some "plan" data in the metadata. 

```
{"compute":{"azEnvironment":"AzurePublicCloud","customData":"","location":"eastus","name":"linux-vm","offer":"manager-server-4-3-llc-image","osType":"Linux","placementGroupId":"","plan":{"name":"gen2","product":"manager-server-4-3-llc-image","publisher":"suse"}
```

3) The resourceURI should be able to be retrieved by getting the `managedBy field` from following the example at the URL above.

 "https://management.azure.com/subscriptions/" + $metadata.compute.subscriptionId + "/resourceGroups/" + $metadata.compute.resourceGroupName + "?api-version=2019-10-01"

but for some reason, even with a managed identity assigned I am getting 

```
{"error":{"code":"AuthorizationFailed","message":"The client 'XXXXXXXXXXXXXX' with object id 'XXXXXXXXXXXXXXXXX' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourceGroups/read' over scope '/subscriptions/XXXXXXXXXXX/resourceGroups/mrg-manager-server-4-3-llc-preview-20231006090809' or the scope is invalid. If access was recently granted, please refresh your credentials."}}
```
This will need to be figured out.

4) A check will have to be added to determine if this is running in local vs k8s and  then use the alternate path to obtains the token, planId and  "resourceId" needed for request body.

```
{
  "resourceId": <guid>, // unique identifier of the resource against which usage is emitted. 
  "quantity": 5.0, // how many units were consumed for the date and hour specified in effectiveStartTime, must be greater than 0 or a double integer
  "dimension": "dim1", // custom dimension identifier
  "effectiveStartTime": "2018-12-01T08:30:14", // time in UTC when the usage event occurred, from now and until 24 hours back
  "planId": "plan1", // id of the plan purchased for the offer
}
```

### PR DESCRIPTION
**Pull request by jesusbv**
This fixes #19

### CODE REVIEW

**Review by smarlowucf — COMMENTED** [review]
Couple thoughts/questions:

- Is there any way to determine vm vs k8s in a deterministic way that doesn't involve an implicit check? There is a chance of env var overlap in a VM that could trigger the wrong code path.
- I wonder if this path we found with the VM is doable in k8s environment? Maybe a question to think about after this PR so as to not hold up vm solutions in the meantime.
- Ideally we'll want to give this code a run through in a VM and a container to ensure there's no regression.

File: csp_billing_adapter_microsoft/plugin.py
Comment (jesusbv) [reply csp_billing_adapter_microsoft/plugin.py:225]: Will fix it, I did not touch initially the path (for k8s) that was working

File: csp_billing_adapter_microsoft/plugin.py
Comment (jesusbv) [reply csp_billing_adapter_microsoft/plugin.py:247]: I thought about it and I did not like the envvar too much but I liked having to carry the config to do the check less
other option is to do the check outside (`meter_billing` level) and pass the param
Now, IMO the responsibility of knowing that should not be of the `meter_biling` though

**Comment by jesusbv** [comment]
> Couple thoughts/questions:
> 
> * Is there any way to determine vm vs k8s in a deterministic way that doesn't involve an implicit check? There is a chance of env var overlap in a VM that could trigger the wrong code path.

I will look into it, I know VM config yaml has `api` key which is not present in any of the NV config files

> * I wonder if this path we found with the VM is doable in k8s environment? Maybe a question to think about after this PR so as to not hold up vm solutions in the meantime.

k8s has access to IMDS AFAIK, not so sure about the managed identity 

> * Ideally we'll want to give this code a run through in a VM and a container to ensure there's no regression.

+1, I have tested it on a VM and it can register the billing and complain about a duplicate billing

File: csp_billing_adapter_microsoft/plugin.py
Comment (KeithMnemonic) [reply csp_billing_adapter_microsoft/plugin.py:247]: you can move the check of ['EXTENSION_RESOURCE_ID'] outside. Perhaps create another function i.e get_node_type() that looks for config['api']and if that is not set (running in k8s) it checks for the 3 required env variables there?

Comment (KeithMnemonic) [reply csp_billing_adapter_microsoft/plugin.py:247]: it could also set them in the function and return them as a tuple or such.

**Comment by KeithMnemonic** [comment]
> > Couple thoughts/questions:
> > 
> > * Is there any way to determine vm vs k8s in a deterministic way that doesn't involve an implicit check? There is a chance of env var overlap in a VM that could trigger the wrong code path.
> 
> I will look into it, I know VM config yaml has `api` key which is not present in any of the NV config files
> 
> > * I wonder if this path we found with the VM is doable in k8s environment? Maybe a question to think about after this PR so as to not hold up vm solutions in the meantime.
> 
> k8s has access to IMDS AFAIK, not so sure about the managed identity
> 
> > * Ideally we'll want to give this code a run through in a VM and a container to ensure there's no regression.
> 
> +1, I have tested it on a VM and it can register the billing and complain about a duplicate billing

Which path are you referring to. @smarlowucf   The docs for the metering API were very specific about how you get the 3 needed values .

EXTENSION_RESOURCE_ID
PLAN_ID
CLIENT_ID

**Comment by smarlowucf** [comment]
> Which path are you referring to.

We found a way to get the required information from metadata in a VM context so I wonder if this would also work in a container context. They are both hitting the same metering API right?

File: csp_billing_adapter_microsoft/plugin.py
Comment (smarlowucf) [reply csp_billing_adapter_microsoft/plugin.py:225]: Fair enough. And that's a good recent to not touch it as well.

**Comment by KeithMnemonic** [comment]
> > Which path are you referring to.
> 
> We found a way to get the required information from metadata in a VM context so I wonder if this would also work in a container context. They are both hitting the same metering API right?

The metadata is different from a VM to a container.  Comparing the values from the env to the metadata that shows up in the logs you see

For k8s, the ENV settings are:

EXTENSION_RESOURCE_ID=/subscriptions/5f40eec9-a9be-4851-90c1-621e6d65df81/resourceGroups/kberger/providers/Microsoft.ContainerService/managedclusters/kberger-nv-test-10-06-2023/providers/Microsoft.KubernetesConfiguration/extensions/mb
PLAN_ID=preview
CLIENT_ID=17d29518-b8c5-4e4b-9b8b-ce035efa4d93

looking at metadata from the logs I see 

'plan': {'name': '', 'product': '', 'publisher': ''},
'resourceId': '/subscriptions/5f40eec9-a9be-4851-90c1-621e6d65df81/resourceGroups/MC_kberger_kberger-nv-test-10-06-2023_eastus/providers/Microsoft.Compute/virtualMachineScaleSets/aks-agentpool-14425402-vmss/virtualMachines/1
'aks-managed-kubeletIdentityClientID', 'value': 'cdfbc815-8228-43bf-b1d0-1ff1461398d3'

I do not see how you can use the same path

File: csp_billing_adapter_microsoft/plugin.py
Comment (KeithMnemonic) [inline csp_billing_adapter_microsoft/plugin.py:79]: you need to fix the corresponding tests

FAIL Required test coverage of 90.0% not reached. Total coverage: 74.84%
=================================================================== short test summary info ===================================================================
FAILED tests/unit/test_plugin.py::test_get_msi_token - TypeError: _get_msi_token() missing 1 required positional argument: 'config'
FAILED tests/unit/test_plugin.py::test_get_msi_token_error - TypeError: _get_msi_token() missing 1 required positional argument: 'config'
FAILED tests/unit/test_plugin.py::test_get_msi_token_invalid - TypeError: _get_msi_token() missing 1 required positional argument: 'config'
================================================================ 3 failed, 14 passed in 0.28s =================================================================

File: csp_billing_adapter_microsoft/plugin.py
Comment (KeithMnemonic) [inline csp_billing_adapter_microsoft/plugin.py:205]: ditto

File: csp_billing_adapter_microsoft/plugin.py
Comment (KeithMnemonic) [inline csp_billing_adapter_microsoft/plugin.py:242]: ditto

**Review by KeithMnemonic — CHANGES_REQUESTED** [review]
some test cases need fixed

File: csp_billing_adapter_microsoft/plugin.py
Comment (smarlowucf) [reply csp_billing_adapter_microsoft/plugin.py:247]: Yeah, for now we could rely on the API option in config. Unless there's a metadata call that's generic that would work across workloads. The benefit is the config is in part of the adapter deployment.

>  IMO the responsibility of knowing that should not be of the meter_billing though

I think it's within the scope of the cloud adapter though. It is a direct connection to billing in terms of "how to bill".

File: csp_billing_adapter_microsoft/plugin.py
Comment (jesusbv) [reply csp_billing_adapter_microsoft/plugin.py:79]: Fixed

**Comment by KeithMnemonic** [comment]
> some test cases need fixed

those pass now but you need to add more to reach the target coverage.

 pytest --cov=csp_billing_adapter_microsoft
===================================================================== test session starts =====================================================================
platform linux -- Python 3.10.13, pytest-7.3.2, pluggy-1.0.0
rootdir: /home/kberger/repos/github/csp/csp-billing-adapter-microsoft
configfile: setup.cfg
testpaths: tests
plugins: cov-4.1.0
collected 17 items

tests/unit/test_plugin.py .................                                                                                                             [100%]

---------- coverage: platform linux, python 3.10.13-final-0 ----------
Name                                        Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------
csp_billing_adapter_microsoft/__init__.py       3      0   100%
csp_billing_adapter_microsoft/plugin.py       156     25    84%   214, 250-256, 280-311, 315-319
-------------------------------------------------------------------------
TOTAL                                         159     25    84%

FAIL Required test coverage of 90.0% not reached. Total coverage: 84.28%

**Review by KeithMnemonic — APPROVED** [review]
LGTM

**Review by smarlowucf — APPROVED** [review]
LGTM, please squash commits before merge.

**Pull request merged**
